### PR TITLE
Add scene graph clone and environment clone

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
@mpowelson I improved the scene graph insertSceneGraph along with add a new command to the environment AddSceneGraphCommand where now every environment starts with one command which is the AddSceneGraphCommand. I will need to create a message for SceneGraph to complete tesseract_ros. Now should have everything in the command history to fully recreate an environment.

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
Also I tried rebasing yours on the latest, but in the end it was easier for me to cherry-pick the commits.

**Comment by Levi-Armstrong** [comment]
Depends on PR #348

File: tesseract/tesseract_environment/include/tesseract_environment/core/environment.h [hunk 1]
```diff
@@ -64,12 +64,64 @@ class Environment
    *
    * Inherited class must call the protected create function
    *
-   * @param scene_graph
-   * @return
+   * @param scene_graph The scene graph to initialize the environment.
+   * @return True if successful, otherwise false
    */
-  virtual bool init(tesseract_scene_graph::SceneGraph::Ptr scene_graph) = 0;
+  template <typename S>
```
Comment (mpowelson) [inline tesseract/tesseract_environment/include/tesseract_environment/core/environment.h:70]: Should the documentation state what should be passed into this template? ie, a state solver?

Comment (mpowelson) [reply tesseract/tesseract_environment/include/tesseract_environment/core/environment.h:70]: Also, do you want to do an explicit instantiation of a KDL state solver (maybe you already do).

**Comment by mpowelson** [comment]
This shouldn't be a breaking change for most people, but the biggest change is that rather than having a custom KDL environment class, the base is now templated by state solver. To migrate, make changes similar to what was done in tesseract.cpp
```
#include <tesseract_environment/kdl/kdl_env.h>

environment_ = std::make_shared<tesseract_environment::KDLEnv>();
environment_->init(scene_graph)
```
becomes
```
#include <tesseract_environment/kdl/kdl_state_solver.h>

environment_ = std::make_shared<tesseract_environment::Environment>();
environment_->init<tesseract_environment::KDLStateSolver>(scene_graph)
```

**Pull request merged**